cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -460,4 +484,58 @@ static PartitionKey parseRowKey(byte[] rowKey, Partitioning partitioning) {
     }
     return builder.build();
   }
+
+  private class BasicPartition implements Partition {
+    protected final String relativePath;
+    protected final PartitionKey key;
+
+    protected BasicPartition(@Nonnull String relativePath, @Nonnull PartitionKey key) {
```

Comment:
I would say we just adapt the "Every parameter is non-null unless explicitly specified" principle so that we don't have a mix of `@Nullable` and `@Nonnull`.